client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,559 @@
+		let _ = self
+			.to_controller
+			.unbounded_send(Action::IncomingConnection(peer_id, incoming_index));
+	}
+
+	/// Notify that connection was dropped (eithere refused or disconnected).
+	pub fn dropped(&self, peer_id: PeerId, reason: DropReason) {
+		let _ = self.to_controller.unbounded_send(Action::Dropped(peer_id, reason));
+	}
+}
+
+/// Direction of a connection
+#[derive(Clone, Copy, Debug)]
+enum Direction {
+	Inbound,
+	Outbound,
+}
+
+/// Status of a connection with a peer.
+#[derive(Clone, Debug)]
+enum PeerState {
+	/// We are connected to the peer.
+	Connected(Direction),
+	/// We are not connected.
+	NotConnected,
+}
+
+impl PeerState {
+	/// Returns true if we are connected with the node.
+	fn is_connected(&self) -> bool {
+		matches!(self, PeerState::Connected(_))
+	}
+}
+
+impl Default for PeerState {
+	fn default() -> PeerState {
+		PeerState::NotConnected
+	}
+}
+
+/// Side of [`ProtocolHandle`] responsible for all the logic. Currently all instances are
+/// owned by [`crate::Peerset`], but they should eventually be moved to corresponding protocols.
+#[derive(Debug)]
+pub struct ProtocolController {
+	/// Set id to use when sending connect/drop requests to `Notifications`.
+	// Will likely be replaced by `ProtocolName` in the future.
+	set_id: SetId,
+	/// Receiver for messages from [`ProtocolHandle`].
+	from_handle: TracingUnboundedReceiver<Action>,
+	/// Number of occupied slots for incoming connections.
+	num_in: u32,
+	/// Number of occupied slots for outgoing connections.
+	num_out: u32,
+	/// Maximum number of slots for incoming connections.
+	max_in: u32,
+	/// Maximum number of slots for outgoing connections.
+	max_out: u32,
+	/// Peers and their connection states (including reserved nodes).
+	nodes: HashMap<PeerId, PeerState>,
+	/// Reserved nodes. Should be always connected and do not occupy peer slots.
+	reserved_nodes: HashMap<PeerId, PeerState>,
+	/// Connect only to reserved nodes.
+	reserved_only: bool,
+	/// Next time to allocate slots. This is done once per second.
+	next_periodic_alloc_slots: Instant,
+	/// Outgoing channel for messages to `Notifications`.
+	to_notifications: TracingUnboundedSender<Message>,
+	/// Peerset handle for checking peer reputation values and getting connection candidates
+	/// with highest reputation.
+	peerset_handle: PeersetHandle,
+}
+
+impl ProtocolController {
+	/// Construct new [`ProtocolController`].
+	pub fn new(
+		set_id: SetId,
+		config: SetConfig,
+		to_notifications: TracingUnboundedSender<Message>,
+		peerset_handle: PeersetHandle,
+	) -> (ProtocolHandle, ProtocolController) {
+		let (to_controller, from_handle) = tracing_unbounded("mpsc_protocol_controller", 10_000);
+		let handle = ProtocolHandle { to_controller };
+		let reserved_nodes = config
+			.reserved_nodes
+			.iter()
+			.map(|p| (*p, PeerState::NotConnected))
+			.collect::<HashMap<PeerId, PeerState>>();
+		// Initialize with bootnodes, but make sure we don't count peers twice.
+		let nodes = config
```

Comment:
All bootnodes are not reserved nodes? This was in the original implementation as well?